File: Problem_1.py
"""
Problem : 1

Time Complexity : O(m*n)
Space Complexity : 
Approach 1 - O(m*n)
Approach 2 - O(n)

Did this code successfully run on Leetcode : Yes
Any problem you faced while coding this : No


"""

# Edit Distance


# Approach - 1
# DP Matrix

class Solution(object):
    def minDistance(self, word1, word2):
        """
        :type word1: str
        :type word2: str
        :rtype: int
        """
        m=len(word1)
        n=len(word2)
        dp=[[0 for _ in range(n+1)] for _ in range(m+1)]
        for i in range(1,n+1):
            dp[0][i]=i
        for i in range(1,m+1):
            dp[i][0]=i
        
        for i in range(1,m+1):
            for j in range(1,n+1):
                if word1[i-1]==word2[j-1]:
                    dp[i][j]=dp[i-1][j-1]
                else:
                    dp[i][j]=1+min(dp[i-1][j],min(dp[i-1][j-1],dp[i][j-1]))

        return dp[m][n]

        
# Approach - 2
# DP array
    
class Solution(object):
    def minDistance(self, word1, word2):
        """
        :type word1: str
        :type word2: str
        :rtype: int
        """
        m=len(word1)
        n=len(word2)
        dp=[0 for _ in range(n+1)]
        for i in range(1,n+1):
            dp[i]=i
        
        for i in range(1,m+1):
            diagUp=dp[0]
            for j in range(0,n+1):
                temp=dp[j]
                if j==0:
                    dp[j]=i
                else:
                    if word1[i-1]==word2[j-1]:
                        dp[j]=diagUp
                    else:
                        dp[j]=1+min(dp[j],min(diagUp,dp[j-1]))
                diagUp=temp

        return dp[n]

File: problem1.cpp
/*
// Time Complexity : O(M*N) m = length of word-1 , n = length of word 2
// Space Complexity : O(M*N)
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this :

Taking the base case for empty string , the base condition 


// Your code here along with comments explaining your approach
Create a dp array of size (len_1+1,len_2+1)
fill the row 0 with increasing range(0,len_2+1)
fill the column 0 with increasing range(0,len_1+1)
for any other matrix element 
insert = back+1
del = top+1
replace = diag_up or diag_up + 1 , 
if the character is equal or not equal
*/

#include<iostream>
#include<vector>
#include<string>

using namespace std;

class Solution {
public:
    int minDistance(string word1, string word2) {
        int len_1 = word1.size();
        int len_2 = word2.size();
        vector<vector<int>> dp_arr(len_1+1,vector<int>(len_2+1,0));
        for(int i{};i<=len_1;++i){
            dp_arr.at(i).at(0) = i;
        }
        for(int j{};j<=len_2;++j){
            dp_arr.at(0).at(j) = j;
        }

        for(int i{1};i<=len_1;++i){
            for(int j{1};j<=len_2;++j){
                int insert = dp_arr.at(i).at(j-1) + 1; //back+1
                int del = dp_arr.at(i-1).at(j)+1; //top+1
                int replace = (word1.at(i-1)!=word2.at(j-1))?dp_arr.at(i-1).at(j-1)+1:dp_arr.at(i-1).at(j-1);
                dp_arr.at(i).at(j) = min(insert,min(del,replace));
            }
        }
        return dp_arr.at(len_1).at(len_2);
    }
};

